README
LSL Toolkit for Images Created by the Prototype All-Sky Imager (PASI)

DESCRIPTION
	This package provides the the PasiImageDB class that is used to
	read images created by the Prototype All-Sky Imager running at 
	LWA1.  This reader supports all three versions of the PASI image
	format found on the LWA1 data archive.

REQUIREMENTS
	python >= 2.6 and python < 3.0
	numpy >= 1.2
	construct >= 2.5.2
	*lsl >= 1.1.0
	*matplotlib >= 0.98.3
	*pyfits >= 3.1
	(* required for some of the scripts)

BUILD
	The PasiImage package is installed as a regular Python package using distutils.  
	Unzip and untar the source distribution. Setup the python interpreter you 
	wish to use for running the package applications and switch to the root of 
	the source distribution tree.
	
	To build the PasiImage package, run:
		'python setup.py build'

TESTING
	To test the as-build PasiImage package, run:
		'python setup.py test'

INSTALL
	Install PasiImage by running:
		'python setup.py install [--prefix=<prefix>|--user]'
		
	If the '--prefix' option is not provided, then the installation 
	tree root directory is the same as for the python interpreter used to run 
	setup.py.  For instance, if the python interpreter is in 
	`/usr/local/bin/python', then <prefix> will be set to '/usr/local'.
	Otherwise, the explicit <prefix> value is taken from the command line
	option.  The package will install files in the following locations:
		<prefix>/bin
		<prefix>/lib/python2.6/site-packages
		<prefix>/share/doc
		<prefix>/share/install
	If an alternate <prefix> value is provided, you should set the PATH
	environment to include directory '<prefix>/bin' and the PYTHONPATH
	environment to include directory '<prefix>/bin' and the PYTHONPATH
	environment to include directory '<prefix>/lib/python2.6/site-packages'.
	
	If the '--user' option is provided, then then installation tree root 
	directory will be in the current user's home directory.

DOCUMENTATION
	See the module's internal documentation.

RELEASE NOTES
	See the CHANGELOG for a detailed list of changes and notes.


CHANGELOG
0.1.4
  * Worked on making the module PEP8 compliant

0.1.3
  * Worked on the WCS values in imagePASI.py and pims2fits.py to make coordinate system more accurate

0.1.2
  * Added a new script to help convert .pims files into FITS images

0.1.1
  * Updated the list of software dependencies
  * Added two scripts to provide examples of how to work with the data

0.1.0
  * Initial release using code from Jake Hartman's PASIFx software


tests/test_pims.py
# -*- coding: utf-8 -*-

"""Unit test for lsl_toolkit.PasiImage module"""

import os
import numpy
import tempfile
import unittest

from lsl_toolkits import PasiImage as PasiImage


__revision__ = "$ Revision: 2 $"
__version__  = "0.1"
__author__    = "Jayce Dowell"


pimsFile = os.path.join(os.path.dirname(__file__), 'data', 'test.pims')


class pims_tests(unittest.TestCase):
	"""A unittest.TestCase collection of unit tests for the lsl_toolkits.PasiImage
	module."""
	
	testPath = None

	def setUp(self):
		"""Turn off all numpy warnings and create the temporary file directory."""

		numpy.seterr(all='ignore')
		self.testPath = tempfile.mkdtemp(prefix='test-pims-', suffix='.tmp')
		
	def test_pims_read(self):
		"""Test reading in an image from a PasiImage file."""

		db = PasiImage.PasiImageDB(pimsFile, 'r')
		
		# Read in the first image with the correct number of elements
		hdr, data, spec = db.readImage()
		## Image
		self.assertEqual(data.shape[0], len(db.header.stokesParams.split(',')))
		self.assertEqual(data.shape[1], db.header.xSize)
		self.assertEqual(data.shape[2], db.header.ySize)
		## Spectra
		self.assertEqual(spec.shape[0], db.header.nSpecChans)
		
		db.close()
		
	def test_pims_loop(self):
		"""Test reading in a collection of images in a loop."""
		
		db = PasiImage.PasiImageDB(pimsFile, 'r')
		
		# Go
		for i,(hdr,data,spec) in enumerate(db):
			i
			
		db.close()
		
	def test_pims_write(self):
		"""Test saving data to the PasiImageDB format."""
		
		# Setup the file names
		testFile = os.path.join(self.testPath, 'test.pims')
		
		db = PasiImage.PasiImageDB(pimsFile, 'r')
		nf = PasiImage.PasiImageDB(testFile, 'w', corrVersion=db.header.corrVersion, 
											imagerVersion=db.header.imagerVersion, 
											station=db.header.station)
											
		# Fill it
		for rec in db:
			nf.addImage(*rec)
			
		# Done
		db.close()
		nf.close()
		
		# Re-open
		db0 = PasiImage.PasiImageDB(pimsFile, 'r')
		db1 = PasiImage.PasiImageDB(testFile, 'r')
		
		# Validate
		## File header
		for attr in ('corrVersion', 'imagerVersion', 'station', 'stokesParams', 'xSize', 'ySize', 'nSpecChans', 'flags'):
			self.assertEqual(getattr(db0.header, attr, None), getattr(db1.header, attr, None))
		for attr in ('xPixelSize', 'yPixelSize', 'startTime', 'stopTime'):
			self.assertAlmostEqual(getattr(db0.header, attr, None), getattr(db1.header, attr, None), 6)
		## First image
		### Image header
		hdr0, img0, spec0 = db0.readImage()
		hdr1, img1, spec1 = db1.readImage()
		for attr in ('visFileName', 'stokesParams'):
			self.assertEqual(getattr(hdr0, attr, None), getattr(hdr1, attr, None))
		for attr in ('startTime', 'centroidTime', 'intLen', 'lst', 'freq', 'bandwidth', 'gain', 'fill', 'zenithRA', 'zenithDec', 'xPixelSize', 'yPixelSize'):
			self.assertAlmostEqual(getattr(hdr0, attr, None), getattr(hdr1, attr, None), 6)
		### Image
		for i in xrange(img0.shape[0]):
			for j in xrange(img0.shape[1]):
				for k in xrange(img0.shape[2]):
					self.assertAlmostEqual(img0[i,j,k], img1[i,j,k], 6)
		### Spectra
		for i in xrange(spec0.shape[0]):
			self.assertAlmostEqual(spec0[i], spec1[i], 6)
			
		db0.close()
		db1.close()
		
	def tearDown(self):
		"""Remove the test path directory and its contents"""
		
		tempFiles = os.listdir(self.testPath)
		for tempFile in tempFiles:
			os.unlink(os.path.join(self.testPath, tempFile))
		os.rmdir(self.testPath)
		self.testPath = None


class pims_test_suite(unittest.TestSuite):
	"""A unittest.TestSuite class which contains all of the lsl_toolkits.PasiImage units 
	tests."""
	
	def __init__(self):
		unittest.TestSuite.__init__(self)
		
		loader = unittest.TestLoader()
		self.addTests(loader.loadTestsFromTestCase(pims_tests)) 


if __name__ == '__main__':
	unittest.main()


scripts/imagePASI.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import numpy
import getopt

from lsl.common.mcs import mjdmpm2datetime
from lsl.common.stations import lwa1
from lsl.sim import vis as simVis
from lsl.imaging import overlay
from lsl import astro

from lsl_toolkits.PasiImage import PasiImageDB

from matplotlib import pyplot as plt
from matplotlib.ticker import NullFormatter


def usage(exitCode=None):
	print """imagePASI.py - Display images in a PASI .pims file

Usage:  imagePASI.py [OPTIONS] file [file [...]]

Options:
-h, --help              Display this help information
-s, --dataset           Data set to image (Default = All)
-n, --no-labels         Disable source and grid labels
-g, --no-grid           Disable the RA/Dec grid
"""
	
	if exitCode is not None:
		sys.exit(exitCode)
	else:
		return True


def parseOptions(args):
	# Build up the configuration
	config = {}
	config['dataset'] = 0
	config['label'] = True
	config['grid'] = True
	config['args'] = []
	
	# Read in and process the command line flags
	try:
		opts, args = getopt.getopt(args, "hs:ng", ["help", "dataset=", "no-labels", "no-grid"])
	except getopt.GetoptError, err:
		# Print help information and exit:
		print str(err) # will print something like "option -a not recognized"
		usage(exitCode=2)
		
	# Work through opts
	for opt, value in opts:
		if opt in ('-h', '--help'):
			usage(exitCode=0)
		elif opt in ('-s', '--dataset'):
			config['dataset'] = int(value)
		elif opt in ('-n', '--no-labels'):
			config['label'] = False
		elif opt in ('-g', '--no-grid'):
			config['grid'] = False
		else:
			assert False
			
	# Add in arguments
	config['args'] = args
	
	# Return configuration
	return config


def main(args):
	config = parseOptions(args)
	filenames = config['args']
	
	# Loop over the input files
	for filename in filenames:
		## Is this file valid?
		try:
			db = PasiImageDB(filename, 'r')
		except Exception as e:
			print "ERROR: %s" % str(e)
			continue
			
		## Setup the array
		if db.header.station == 'LWA1':
			aa = simVis.buildSimArray(lwa1, lwa1.getAntennas()[0::2], numpy.array([38e6,]))
		else:
			aa = None
			
		## Go!
		for i,(hdr,img,spec) in enumerate(db):
			if config['dataset'] != 0 and config['dataset'] != (i+1):
				continue
				
			mjd = int(hdr.startTime)
			mpm = int((hdr.startTime - mjd)*86400*1000.0)
			tStart = mjdmpm2datetime(mjd, mpm)
			if aa is not None:
				aa.set_jultime(hdr.centroidTime + astro.MJD_OFFSET)
				
			stokes = hdr.stokesParams.split(',')
			
			### Save the image size for later
			imSize = img.shape[-1]
			
			### Zero outside of the horizon so avoid problems
			pCntr = imSize/2 + 1 + 0.5 * ((imSize+1)%2)
			pScale = hdr['xPixelSize']
			sRad   = 360.0/pScale/numpy.pi / 2
			x = numpy.arange(1, img.shape[-2]+1, dtype=numpy.float32) - pCntr
			y = numpy.arange(1, img.shape[-1]+1, dtype=numpy.float32) - pCntr
			x /= -sRad
			y /= sRad
			x,y = numpy.meshgrid(x,y)
			invalid = numpy.where( (x**2 + y**2) > 1 )
			img[:, invalid[0], invalid[1]] = 0.0
			
			### Try and set the image scale correctly for the display
			x2 = x - 1 / sRad
			y2 = y - 1 / sRad
			extent = (x2.max(), x2.min(), y.min(), y.max())
			
			### Loop over Stokes parameters
			fig = plt.figure()
			for j,label in enumerate(stokes):
				ax = fig.add_subplot(2, 2, j+1)
				ax.imshow(img[j,:,:].T, origin='lower', interpolation='nearest', extent=extent)
				ax.set_title(label)
				ax.set_xlim((1,-1))
				ax.set_ylim((-1,1))
				
				## Turn off tick marks
				ax.xaxis.set_major_formatter( NullFormatter() )
				ax.yaxis.set_major_formatter( NullFormatter() )
				
				## Is we know where we are, overlay some stuff
				if aa is not None:
					# Horizon
					overlay.horizon(ax, aa)
					# RA/Dec graticle
					if config['grid']:
						overlay.graticleRADec(ax, aa)
					# Source positions
					overlay.sources(ax, aa, simVis.srcs, label=config['label'])
					
			fig.suptitle('%.3f MHz @ %s' % (hdr.freq/1e6, tStart.strftime("%Y/%m/%d %H:%M:%S")))
			plt.show()
			
		## Done
		db.close()


if __name__ == "__main__":
	numpy.seterr(all='ignore')
	main(sys.argv[1:])
	


scripts/summarizePASIImage.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import getopt

from lsl.common.mcs import mjdmpm2datetime

from lsl_toolkits.PasiImage import PasiImageDB


def usage(exitCode=None):
	print """summarizePASIImage.py - Print metadata about a PASI .pims file

Usage:  summarizePASIImage.py [OPTIONS] file [file [...]]

Options:
-h, --help              Display this help information
"""
	
	if exitCode is not None:
		sys.exit(exitCode)
	else:
		return True


def parseOptions(args):
	# Build up the configuration
	config = {}
	
	# Read in and process the command line flags
	try:
		opts, args = getopt.getopt(args, "h", ["help",])
	except getopt.GetoptError, err:
		# Print help information and exit:
		print str(err) # will print something like "option -a not recognized"
		usage(exitCode=2)
		
	# Work through opts
	for opt, value in opts:
		if opt in ('-h', '--help'):
			usage(exitCode=0)
		else:
			assert False
			
	# Add in arguments
	config['args'] = args
	
	# Return configuration
	return config


def main(args):
	config = parseOptions(args)
	filenames = config['args']
	
	# Loop over the input files
	for filename in filenames:
		## Is this file valid?
		try:
			db = PasiImageDB(filename, 'r')
		except Exception as e:
			print "ERROR: %s" % str(e)
			continue
			
		## Report - overall
		print "Filename: %s" % os.path.basename(filename)
		print "  Correlator: %s" % db.header.corrVersion
		print "  Imager: %s" % db.header.imagerVersion
		print "  Station: %s" % db.header.station
		print "  Stokes Parameters: %s" % db.header.stokesParams
		print "  Image Size: %i by %i with %.3f deg/px" % (db.header.xSize, db.header.ySize, db.header.xPixelSize)
		print "  Number of Integrations: %i" % db.nIntegrations
		
		## Report - first image
		db.seek(0)
		hdr, data, spc = db.readImage()
		mjd = int(hdr.startTime)
		mpm = int((hdr.startTime - mjd)*86400*1000.0)
		tStart = mjdmpm2datetime(mjd, mpm)
		print "    First Image:"
		print "      Start Time: %s" % tStart.strftime("%Y/%m/%d %H:%M:%S.%f")
		print "      Integration Time: %.3f s" % (hdr.intLen*86400.0,)
		print "      Tuning: %.3f MHz" % (hdr.freq/1e6,)
		print "      Bandwidth: %.3f kHz" % (hdr.bandwidth/1e3,)
		
		## Report - last image
		db.seek(db.nIntegrations-1)
		hdr, data, spc = db.readImage()
		mjd = int(hdr.startTime)
		mpm = int((hdr.startTime - mjd)*86400*1000.0)
		tStart = mjdmpm2datetime(mjd, mpm)
		print "    Last Image:"
		print "      Start Time: %s" % tStart.strftime("%Y/%m/%d %H:%M:%S.%f")
		print "      Integration Time: %.3f s" % (hdr.intLen*86400.0,)
		print "      Tuning: %.3f MHz" % (hdr.freq/1e6,)
		print "      Bandwidth: %.3f kHz" % (hdr.bandwidth/1e3,)
		
		## Done
		print " "
		db.close()


if __name__ == "__main__":
	main(sys.argv[1:])
	
	


scripts/pims2fits.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import numpy
import getopt
import pyfits
from datetime import datetime, timedelta

from lsl.common.mcs import mjdmpm2datetime

from lsl_toolkits.PasiImage import PasiImageDB


def usage(exitCode=None):
	print """pims2fits.py - Convert the images contained in one or more .pims
files into FITS images.

Usage:  pims2fits.py [OPTIONS] file [file [...]]

Options:
-h, --help              Display this help information
-f, --force             Force overwriting of FITS files
-v, --verbose           Be verbose during the conversion
"""
	
	if exitCode is not None:
		sys.exit(exitCode)
	else:
		return True


def parseOptions(args):
	# Build up the configuration
	config = {}
	config['force'] = False
	config['verbose'] = False
	config['args'] = []
	
	# Read in and process the command line flags
	try:
		opts, args = getopt.getopt(args, "hfv", ["help", "force", "verbose"])
	except getopt.GetoptError, err:
		# Print help information and exit:
		print str(err) # will print something like "option -a not recognized"
		usage(exitCode=2)
		
	# Work through opts
	for opt, value in opts:
		if opt in ('-h', '--help'):
			usage(exitCode=0)
		elif opt in ('-f', '--force'):
			config['force'] = True
		elif opt in ('-v', '--verbose'):
			config['verbose'] = True
		else:
			assert False
			
	# Add in arguments
	config['args'] = args
	
	# Return configuration
	return config


def main(args):
	# Parse the command line
	config = parseOptions(args)
	filenames = config['args']
	
	# Loop over input .pims files
	for filename in filenames:
		print "Working on '%s'..." % os.path.basename(filename)
		
		## Open the image database
		try:
			db = PasiImageDB(filename, mode='r')
		except Exception as e:
			print "ERROR: %s" % str(e)
			continue
			
		##  Loop over the images contained in it
		fitsCounter = 0
		for i,(header,data,spec) in enumerate(db):
			if config['verbose']:
				print "  working on integration #%i" % (i+1)
				
			## Reverse the axis order so we can get it right in the FITS file
			data = numpy.transpose(data, [0,2,1])
			
			## Save the image size for later
			imSize = data.shape[-1]
			
			## Zero outside of the horizon so avoid problems
			pScale = header['xPixelSize']
			sRad   = 360.0/pScale/numpy.pi / 2
			x = numpy.arange(data.shape[-2]) - 0.5
			y = numpy.arange(data.shape[-1]) - 0.5
			x,y = numpy.meshgrid(x,y)
			invalid = numpy.where( ((x-imSize/2.0)**2 + (y-imSize/2.0)**2) > (sRad**2) )
			data[:, invalid[0], invalid[1]] = 0.0
			ext = imSize/(2*sRad)
			
			## Convert the start MJD into a datetime instance and then use 
			## that to come up with a stop time
			mjd = int(header['startTime'])
			mpm = int((header['startTime'] - mjd)*86400.0*1000.0)
			tInt = header['intLen']*86400.0
			dateObs = mjdmpm2datetime(mjd, mpm)
			dateEnd = dateObs + timedelta(seconds=int(tInt), microseconds=int((tInt-int(tInt))*1000000))
			if config['verbose']:
				print "    start time: %s" % dateObs
				print "    end time: %s" % dateEnd
				print "    integration time: %.3f s" % tInt
				print "    frequency: %.3f MHz" % header['freq']
				
			## Create the FITS HDU and fill in the header information
			hdu = pyfits.PrimaryHDU(data=data)
			hdu.header['TELESCOP'] = 'LWA1'
			### Date and time
			hdu.header['DATE-OBS'] = dateObs.strftime("%Y-%m-%dT%H:%M:%S")
			hdu.header['END_UTC'] = dateEnd.strftime("%Y-%m-%dT%H:%M:%S")
			hdu.header['EXPTIME'] = tInt
			### Coordinates - sky
			hdu.header['CTYPE1'] = 'RA---SIN'
			hdu.header['CRPIX1'] = imSize/2 + 1 + 0.5 * ((imSize+1)%2)
			hdu.header['CDELT1'] = -360.0/(2*sRad)/numpy.pi
			hdu.header['CRVAL1'] = header['zenithRA']
			hdu.header['CUNIT1'] = 'deg'
			hdu.header['CTYPE2'] = 'DEC--SIN'
			hdu.header['CRPIX2'] = imSize/2 + 1 + 0.5 * ((imSize+1)%2)
			hdu.header['CDELT2'] = 360.0/(2*sRad)/numpy.pi
			hdu.header['CRVAL2'] = header['zenithDec']
			hdu.header['CUNIT2'] = 'deg'
			### Coordinates - Stokes parameters
			hdu.header['CTYPE3'] = 'STOKES'
			hdu.header['CRPIX3'] = 1
			hdu.header['CDELT3'] = 1
			hdu.header['CRVAL3'] = 1
			hdu.header['LONPOLE'] = 180.0
			hdu.header['LATPOLE'] = 90.0
			### LWA1 approximate beam size
			beamSize = 2.2*74e6/header['freq']
			hdu.header['BMAJ'] = beamSize/header['xPixelSize']
			hdu.header['BMIN'] = beamSize/header['xPixelSize']
			hdu.header['BPA'] = 0.0
			### Frequency
			hdu.header['RESTFREQ'] = header['freq']
			
			## Write it to disk
			outName = "pasi_%.3fMHz_%s.fits" % (header['freq']/1e6, dateObs.strftime("%Y-%m-%dT%H-%M-%S"))
			hdulist = pyfits.HDUList([hdu,])
			hdulist.writeto(outName, clobber=config['force'])
			
			## Update the counter
			fitsCounter += 1
			
		## Done with this collection
		db.close()
		
		## Report
		print "-> wrote %i FITS files" % fitsCounter


if __name__ == "__main__":
	main(sys.argv[1:])
	
